rasa_core/policies/two_stage_fallback.py

```diff
@@ -0,0 +1,168 @@
+            if not should_fallback:
+                logger.debug("User clarified instead of confirming.")
+                result = confidence_scores_for(ACTION_REVERT_FALLBACK_EVENTS,
+                                               FALLBACK_SCORE, domain)
+            else:
+                result = confidence_scores_for(self.fallback_action_name,
+                                               FALLBACK_SCORE, domain)
+        elif should_fallback:
+            logger.debug("User has to confirm intent.")
+            result = confidence_scores_for(ACTION_DEFAULT_ASK_CONFIRMATION,
+                                           FALLBACK_SCORE,
+                                           domain)
+        else:
+            result = self.fallback_scores(domain, self.core_threshold)
+
+        return result
+
+    def __is_user_input_expected(self, tracker: DialogueStateTracker) -> bool:
+        return tracker.latest_action_name in [ACTION_DEFAULT_ASK_CONFIRMATION,
+                                              ACTION_DEFAULT_ASK_CLARIFICATION,
+                                              self.fallback_action_name]
+
+    def __results_for_user_denied(self, tracker: DialogueStateTracker,
+                                  domain: Domain) -> List[float]:
+        has_denied_before = tracker.last_executed_has(
+            ACTION_DEFAULT_ASK_CLARIFICATION,
+            skip=1)
+
+        if has_denied_before:
+            return confidence_scores_for(self.fallback_action_name,
+                                         FALLBACK_SCORE, domain)
+        else:
+            return confidence_scores_for(ACTION_DEFAULT_ASK_CLARIFICATION,
+                                         FALLBACK_SCORE, domain)
+
+    def persist(self, path: Text) -> None:
+        """Persists the policy to storage."""
+        config_file = os.path.join(path, 'two_stage_fallback_policy.json')
+        meta = {
+            "nlu_threshold": self.nlu_threshold,
+            "core_threshold": self.core_threshold,
+            "fallback_action_name": self.fallback_action_name,
+        }
+        utils.create_dir_for_file(config_file)
+        utils.dump_obj_as_json_to_file(config_file, meta)
+
+    @classmethod
+    def load(cls, path: Text) -> 'FallbackPolicy':
+        meta = {}
+        if os.path.exists(path):
+            meta_path = os.path.join(path, "two_stage_fallback_policy.json")
+            if os.path.isfile(meta_path):
+                meta = json.loads(utils.read_file(meta_path))
+
+        return cls(**meta)
+
+
+def has_user_confirmed(last_intent: Text,
+                       tracker: DialogueStateTracker) -> bool:
+    return tracker.last_executed_has(name=ACTION_DEFAULT_ASK_CONFIRMATION) \
+        and last_intent == USER_INTENT_CONFIRM
+
+
+def _has_user_denied(last_intent: Text,
+                     tracker: DialogueStateTracker) -> bool:
+    return tracker.last_executed_has(name=ACTION_DEFAULT_ASK_CONFIRMATION) \
```

Comment:
this should probably use `has_user_clarified `

Reply:
what do you means by this?

Reply:
@tmbo